lib/graphql/query/serial_execution/value_resolution.rb

```diff
@@ -80,7 +80,9 @@ def result
             wrapped_type = field_type.of_type
             strategy_class = get_strategy_for_kind(wrapped_type.kind)
             inner_strategy = strategy_class.new(value, wrapped_type, target, parent_type, ast_field, execution_context)
-            inner_strategy.result
+            value = inner_strategy.result
+            raise GraphQL::InvalidNullError.new(ast_field.name, nil) if value.nil?
```

Comment:
The [Scalars Result Coercion](http://facebook.github.io/graphql/#sec-Scalars) section of the spec it says:

> Since this coercion behavior is not observable to clients of the GraphQL server, the precise rules of coercion are left to the implementation. The only requirement is that the server must yield values which adhere to the expected Scalar type.

so the graphql spec doesn't say what the correct behaviour is, just that the result can be either coerced if it can be reasonably coerced or a field error must be produced.

The [Error handling](http://facebook.github.io/graphql/#sec-Error-handling) section is clear about field error resulting in the error being added to the "errors" list though

> If an error occurs when resolving a field, it should be treated as though the field returned null, and an error must be added to the "errors" list in the response.

In this case I don't think we want the type coercion to silently change semantics.  I think that should be considered a bug and reported as an internal error for fixing.  If the returning `nil` is appropriate in a field for some garbage value, then we should be explicit about that and return `nil` in the field resolver.